internals/cli/masker/writer.go

```diff
@@ -1,225 +1,171 @@
-// So after the third f is inserted, the currentIndex is decreased by 3 with the following code.
-func (m *sequenceMatcher) findShift() int {
-	for offset := 1; offset <= m.currentIndex; offset++ {
-		ok := true
-		for i := 0; i < m.currentIndex-offset; i++ {
-			if m.sequence[i] != m.sequence[i+offset] {
-				ok = false
-				break
+func (s *stream) lookForMatches(n int) {
+	searchBuf := s.buf.Bytes()
+	bufLen := n + s.maxLookback
+	if bufLen > len(searchBuf) {
+		bufLen = len(searchBuf)
+	}
+	searchBuf = searchBuf[:bufLen]
+	for _, seq := range s.masker.MatchSequences {
+
+		// finding matches with a regexp is really easy but it is far from efficient.
+		for _, match := range seq.FindAllIndex(searchBuf, -1) {
+			for i := match[0]; i < match[1]; i++ {
+				s.matches[s.index+int64(i)] = struct{}{}
 			}
 		}
-		if ok {
-			return offset
-		}
 	}
-	return m.currentIndex
 }
 
-// InProgress returns whether this sequenceMatcher is currently partially matching.
-//
-// For example, if the sequence is "foobar" and the registered input is "foob", InProgress() returns true.
-func (m *sequenceMatcher) InProgress() bool {
-	return m.currentIndex > 0
+type trigger struct {
+	length int
+	stream *stream
+	timer  *time.Timer
 }
 
-// Reset forgets the current match.
-func (m *sequenceMatcher) Reset() {
-	m.currentIndex = 0
-}
-
-// maskByte represents a byte and whether the byte should be masked or not.
-type maskByte struct {
-	byte
-	masked bool
-}
+type Masker struct {
+	BufferDelay    time.Duration
+	MatchSequences []*regexp.Regexp
 
-// MaskedWriter wraps an io.Writer which masks all occurrences of masks by maskString.
-// If no write is made for timeout on the io.Writer, any matches in progress are reset
-// and the buffer is flushed. This is to ensure that the writer does not hang on partial matches.
-type MaskedWriter struct {
-	w          io.Writer
-	maskString string
-	matchers   []matcher
-	timeout    time.Duration
-
-	buf             []maskByte
-	incomingBytesCh chan []byte
-	outputTimeoutCh chan struct{}
-	outputCh        chan []maskByte
-	errCh           chan error
-	wg              sync.WaitGroup
+	lock         sync.Mutex
+	triggers     []trigger
+	flushAllChan chan struct{}
 }
 
-// NewMaskedWriter returns a new MaskedWriter that masks all occurrences of sequences in masks with maskString.
-func NewMaskedWriter(w io.Writer, masks [][]byte, maskString string, timeout time.Duration) *MaskedWriter {
-	matchers := make([]matcher, len(masks))
-	for i, mask := range masks {
-		matchers[i] = &sequenceMatcher{
-			sequence: mask,
-		}
-	}
-	return &MaskedWriter{
-		w:               w,
-		maskString:      maskString,
-		matchers:        matchers,
-		timeout:         timeout,
-		errCh:           make(chan error, 1),
-		outputTimeoutCh: make(chan struct{}, 1),
-		incomingBytesCh: make(chan []byte, 256),
-		outputCh:        make(chan []maskByte),
+func (m *Masker) AddStream(w io.Writer) io.Writer {
```

Comment:
Had a quick look, one think I noticed is that it looks like the `AddStream` function assumes the source (`io.Reader`) and destination (`io.Writer`) are the same. Consider whether you want to make that assumption or make both configurable.

Reply:
I don't understand what you mean here :thinking: Afaik there is no `io.Reader` involved here. Which reader are you referring to?

Reply:
Ahhh okay, I understood it wrong I think. I thought you read from the thingy you pass to the child process as `stdout` and `stderr` right, but that is actually a custom writer not a reader. The custom writer is passed to the child process and then on every write to the custom writer it writes to the underlying writer (`os.StdOut` or `os.StdErr`) with modifications like masking. 

Consider my comment unsaid. 😊

Reply:
What comment?